File: data-structures/graph/typescript/adjacency-matrix-graph.ts
class AdjacencyMatrixGraph {
	private vertices: string[];
	private matrix: number[][];

	constructor() {
		this.vertices = [];
		this.matrix = [];
	}

	addVertex(vertex: string): void {
		if (this.vertices.includes(vertex)) {
			throw new Error('Vertex already exists');
		}

		this.vertices.push(vertex);

		const newSize = this.vertices.length;
		this.matrix.forEach(row => row.push(0));
		this.matrix.push(new Array(newSize).fill(0));
	}

	addEdge(vertex1: string, vertex2: string): void {
		const index1 = this.vertices.indexOf(vertex1);
		const index2 = this.vertices.indexOf(vertex2);

		if (index1 === -1 || index2 === -1) {
			throw new Error('Vertex not found in the graph');
		}

		this.matrix[index1][index2] = 1;
		this.matrix[index2][index1] = 1;
	}

	removeVertex(vertex: string): void {
		const index = this.vertices.indexOf(vertex);
		if (index === -1) {
			throw new Error('Vertex not found in the graph');
		}

		this.vertices.splice(index, 1);

		this.matrix.splice(index, 1);
		this.matrix.forEach(row => row.splice(index, 1));
	}

	removeEdge(vertex1: string, vertex2: string): void {
		const index1 = this.vertices.indexOf(vertex1);
		const index2 = this.vertices.indexOf(vertex2);

		if (index1 === -1 || index2 === -1) {
			throw new Error('Vertex not found in the graph');
		}

		this.matrix[index1][index2] = 0;
		this.matrix[index2][index1] = 0;
	}
}

const matrixGraph = new AdjacencyMatrixGraph();

matrixGraph.addVertex('Tokyo');
matrixGraph.addVertex('Toronto');
matrixGraph.addVertex('Rio');
matrixGraph.addEdge('Tokyo', 'Toronto');
matrixGraph.addEdge('Tokyo', 'Rio');
matrixGraph.removeEdge('Tokyo', 'Rio');

console.log(matrixGraph);

File: data-structures/graph/python/adjacency_matrix_graph.py
from typing import List


class AdjacencyMatrixGraph:
    def __init__(self):
        self.vertices: List[str] = []
        self.matrix: List[List[int]] = []

    def add_vertex(self, vertex: str) -> None:
        if vertex in self.vertices:
            raise ValueError('Vertex already exists')

        self.vertices.append(vertex)
        new_size = len(self.vertices)
        for row in self.matrix:
            row.append(0)
        self.matrix.append([0] * new_size)

    def add_edge(self, vertex1: str, vertex2: str) -> None:
        index1 = self.vertices.index(vertex1)
        index2 = self.vertices.index(vertex2)

        self.matrix[index1][index2] = 1
        self.matrix[index2][index1] = 1

    def remove_vertex(self, vertex: str) -> None:
        index = self.vertices.index(vertex)
        self.vertices.pop(index)
        self.matrix.pop(index)
        for row in self.matrix:
            row.pop(index)

    def remove_edge(self, vertex1: str, vertex2: str) -> None:
        index1 = self.vertices.index(vertex1)
        index2 = self.vertices.index(vertex2)

        self.matrix[index1][index2] = 0
        self.matrix[index2][index1] = 0
